### ISSUE TITLE
Rule proposal: Flag probably incorrect usage of unique symbols

### ISSUE THREAD

**Issue by wdanilo** [typescript-eslint/typescript-eslint#8309]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

TS uses structural typing, which is cool, but sometimes it makes catching type errors extremely hard. TS introduced unique types, that ALMOST allow us creating "new types". Consider the following code:

```ts
export type Id = string & {readonly __tag: unique symbol}
export const Id = (val: string): Id => val as Id

function test1(arg: string) {}
function test2(arg: Id) {}

test1("id") // Currently OK, which is a good behavior.
test1(Id("id")) // Currently OK, should FAIL.

test2("id") // Currently FAILS, which is a good behavior.
test2(Id("id")) // Currently OK, which is a good behavior.
```

In this code, we define a "new type" `Id`, which we can use to distinguish between string usage as "ID" and other string usages. Unfortunately, not all incorrect usages are caught, as we can still pass `Id("something")` to a function expecting a string. This proposal is about a lint rule that will flag such places.

### Fail Cases

```TypeScript
export type Id = string & {readonly __tag: unique symbol}
export const Id = (val: string): Id => val as Id

function test1(arg: string) {}
test1(Id("id"))
```

### Pass Cases

```TypeScript
export type Id = string & {readonly __tag: unique symbol}
export const Id = (val: string): Id => val as Id

function test1(arg: string) {}
function test2(arg: Id) {}

test1("id")
test2(Id("id"))
```

**Comment by bradzacher** [comment]
> as we can still pass Id("something") to a function expecting a string

This is expected for the vast, vast majority of cases - you don't want to treat the nominal type as truly unique - you want to treat it as an extension of its parent. Otherwise you technically shouldn't do things like `id.length` etc.

For example we have a type for a string that's been validated as a path. We are more than happy to be able to do something like:

```ts
function join(a: string, b: string): Path { ... }
join(Path('/'), 'foo/bar');
```

For many usecases there's no reason we wouldn't want to be able to treat the nominal as its parent type.

---

That aside - there's also no truly standardised way of declaring a nominal type - meaning we would need to do a heuristic-based approach to guess if a type is meant to be considered a nominal type.

Additionally such a rule would be prohibitively expensive - we would need to look at every "assignment" - so all call expression arguments, every assignment expression, every variable declaration, every return statement, every object property assignment, every class property default value assignment, every JSX attribute assignment, etc ... That's a heck of a lot of checks for cases where we need to do loose heuristic checks on.

I definitely get what you're going for - you're looking for a way to do [opaque types without a constraint](https://flow.org/en/docs/types/opaque-types/) from flow.

But it's a super super super niche usecase to want that exact behaviour on top of the niche usecase of nominal types - so that means this proposal fails the 3rd checkbox above. In addition to the expensive nature of the rule - this proposal is a non-starter within our project.

That doesn't mean you can't [build on top of our tooling](https://typescript-eslint.io/developers/custom-rules) to create your own plugin! We just don't believe it belongs in this project.

**Comment by wdanilo** [comment]
@bradzacher Gotcha, thanks a lot @bradzacher for the in-depth explanation. I understand all of your points, and as someone coming from Haskell and Rust, this makes me sad, as I always feel that my scripts in TS are not properly type-secure. Anyway, I have to live with it. Thanks again :)

**Comment by wdanilo** [comment]
Out of curiosity, would the implementation still be so hard if we for example create a rule to flag such usages only for types that have the `__newtype` field typed as `unique symbol` (`__newtype: unique symbol`)? This would give control to users who want to opt-in and would apply only to these specific cases.

**Comment by bradzacher** [comment]
That would remove the ambiguity as there would no longer be a heuristic - but it would not remove scale of the rule.

As an example - there are 3 rules which do a similar check, but for `any` - `no-unsafe-argument`, `no-unsafe-assignment`, `no-unsafe-return`. Together they account for around 1,000 lines - which is a lot of code (not including tests either).
